python_modules/dagster/dagster/_core/definitions/definitions_class.py

```diff
@@ -723,20 +724,32 @@ def get_all_asset_specs(self) -> Sequence[AssetSpec]:
         return [asset_node.to_asset_spec() for asset_node in asset_graph.asset_nodes]
 
     @experimental
-    def with_reconstruction_metadata(self, state_metadata: Mapping[str, Any]) -> Self:
-        """Add metadata to the Definitions object. This is typically used to cache data
+    def with_reconstruction_metadata(self, reconstruction_metadata: Mapping[str, str]) -> Self:
+        """Add reconstruction metadata to the Definitions object. This is typically used to cache data
         loaded from some external API that is computed during initialization of a code server.
         The cached data is then made available on the DefinitionsLoadContext during
         reconstruction of the same code location context (such as a run worker), allowing use of the
-        cached data to avoid additional external API queries. Values must be JSON-serializable.
+        cached data to avoid additional external API queries. Values are expected to be serialized
+        in advance and must be strings.
         """
-        state_metadata = {
-            k: CodeLocationReconstructionMetadataValue(v) for k, v in state_metadata.items()
+        try:
+            check.mapping_param(
+                reconstruction_metadata, "reconstruction_metadata", key_type=str, value_type=str
+            )
+        except CheckError:
+            raise DagsterInvariantViolationError(
+                "Reconstruction metadata must be a Mapping[str, str]."
+                " State-representing values are expected to be serialized"
+                " before being passed to this method. Got:\n\n{reconstruction_metadata}"
+            )
+        normalized_metadata = {
+            k: CodeLocationReconstructionMetadataValue(v)
+            for k, v in reconstruction_metadata.items()
```

Comment:
nit: would prefer to add an `is_mapping` to check rather than rely on exception as control flow